File: bfs.cpp
/// use ...\> g++ bfs.cpp -std=c++11 for compiling.
#include<bits/stdc++.h>
using namespace std;
#define rep(i,n,m) for(int i = n;i < m;i++)
#define pb push_back
typedef long long int ll;
#define max(a,b) a >= b ? a:b
#define min(a,b) a >= b ? b:a
#define F first
#define S second
#define N 1000000007
int main()
{
    ll v,k,n,m;cin>>v>>k;
    // Initializing a graph 
    vector <ll> adj[v];
    rep(i,0,k)
    {
        cin>>n>>m;
        n--;m--;
        adj[n].pb(m);
        adj[m].pb(n);
    }    
    rep(i,0,k)
    {
        cout<<i<<':'<<' ';
        rep(j,0,adj[i].size()) cout<<adj[i][j]<<' ';
        cout<<endl;
    }
    
    ll start = 0;
    unordered_map <ll,ll> parent;
    unordered_map <ll,ll> level;
    vector<ll> frontier;
    frontier.pb(start);
    parent[start] = -1;
    level[start] = 0;
    ll i = 1;
    // Doing a BFS
    while (frontier.empty() == false)
    {
        vector <ll> n;
        rep(u,0,frontier.size())
        {
            rep(v,0,adj[frontier[u]].size())
            {
                if (level.find(adj[frontier[u]][v]) == level.end())
                {
                    level[adj[frontier[u]][v]] = i;
                    parent[adj[frontier[u]][v]] = frontier[u];
                    n.pb(adj[frontier[u]][v]);
                }
            }
        }
        frontier = n;
        i++;
    }
    for(auto & x:parent)
    {
        cout<<x.F <<" --> "<<x.S<<endl;
    }
    cout<<endl;
    for(auto & x:level)
    {
        cout<<x.F <<" --> "<<x.S<<endl;
    }
    return 0;
}

File: bfs.py
adj = {}
k = input()
graph = input()
node = input()
vertex = []
if graph == 0:
	for i in range(k):
		n1,m1 = raw_input().split()
		n,m = int(n1),int(m1)
		if adj.has_key(n):
			adj[n].append(m)
		if adj.has_key(m):
			adj[m].append(n)
		if adj.has_key(n) is False:
			adj[n] = [m]
		if adj.has_key(m) is False:
			adj[m] = [n]

if graph == 1:
	for i in range(k):
		n1,m1 = raw_input().split()
		n,m = int(n1),int(m1)
		if adj.has_key(n):
			adj[n].append(m)
		else:
			adj[n] = [m]

for key in adj.keys():
    vertex.append(key)
print adj
"""
	Doing BFS
"""
start = 1
level,parent = {},{}
level[start] = 0
parent[start] = None
i = 1
frontier = [start]
deepest = 1
while frontier:
	new = []
	for u in frontier:
		for v in adj[u]:
			if level.has_key(v) is False:
				level[v] = i
				parent[v] = u
				new.append(v)
				deepest = v
	frontier = new
	i += 1

#print deepest,level,parent

"""
	Printing the shortest path between start and node.
"""
def shortest_path(node,start):
    path = []
    while node != start:
    	path.append(str(node))
    	node = parent[node]
    path.append(str(start))
    path.reverse()
    print " --> ".join(path)
    return len(path)-1
"""
    Finding the diameter of a tree
"""
start = deepest
level,parent = {},{}
level[start] = 0
parent[start] = None
i = 1
frontier = [start]
while frontier:
	new = []
	for u in frontier:
		for v in adj[u]:
			if level.has_key(v) is False:
				level[v] = i
				parent[v] = u
				new.append(v)
				farthest = v
	frontier = new
	i += 1
print level,parent
print shortest_path(farthest,deepest)
